ui/lib/replication/addon/components/enable-replication-form.js

```diff
@@ -0,0 +1,126 @@
+ * <EnableReplicationForm @replicationMode="dr" @canEnablePrimary={{true}} @canEnableSecondary={{false}} @performanceReplicationDisabled={{false}} @onSuccess={{this.reloadCluster}} />
+ *    @param {string} replicationMode - should be one of "dr" or "performance"
+ *    @param {boolean} canEnablePrimary - if the capabilities allow the user to enable a primary cluster
+ *    @param {boolean} canEnableSecondary - if the capabilities allow the user to enable a secondary cluster
+ *    @param {boolean} performanceMode - should be "primary", "secondary", or "disabled". If enabled, form will show a warning when attempting to enable DR secondary
+ *    @param {Promise} onSuccess - (optional) callback called after successful replication enablement. Must be a promise.
+ *    @param {boolean} doTransition - (optional) if provided, passed to onSuccess callback to determine if a transition should be done
+ *  />
+ * ```
+ */
+export default class EnableReplicationFormComponent extends Component {
+  @service version;
+  @service store;
+
+  @tracked error = '';
+  @tracked showExplanation = false;
+  data = new EnablePayload();
+
+  get performanceReplicationEnabled() {
+    return this.args.performanceMode !== 'disabled';
+  }
+
+  get tokenIncludesAPIAddr() {
+    const config = decodeConfigFromJwt(this.token);
+    return config && config.addr ? true : false;
+  }
+
+  get disallowEnable() {
+    if (this.args.replicationMode === 'performance' && this.version.hasPerfReplication === false) {
+      return true;
+    }
+    const { mode, tokenIncludesAPIAddr, primary_api_addr } = this.data;
+    if (mode !== 'secondary' || tokenIncludesAPIAddr || (!tokenIncludesAPIAddr && primary_api_addr)) {
+      return false;
+    }
+    return true;
+  }
+
+  async onSuccess(resp, clusterMode) {
+    // clear form
+    this.data.reset();
+    // call callback
+    if (this.args.onSuccess) {
+      await this.args.onSuccess(resp, this.args.replicationMode, clusterMode, this.args.doTransition);
+    }
+  }
+
+  @action inputChange(evt) {
+    const name = evt.target.name;
+    const val = evt.target.value;
+    this.data[name] = val;
+  }
+
+  @task
+  @waitFor
+  *enableReplication(replicationMode, clusterMode, data) {
+    const payload = data.allKeys.reduce((newData, key) => {
+      var val = data[key];
+      if (isPresent(val)) {
+        newData[key] = val;
+      }
+      return newData;
+    }, {});
+    delete payload.mode;
+    try {
+      const resp = yield this.store
+        .adapterFor('cluster')
+        .replicationAction('enable', replicationMode, clusterMode, payload);
+      yield this.onSuccess(resp, clusterMode);
+    } catch (e) {
+      this.error = errorMessage(e, 'Enable replication failed. Check Vault logs for details.');
+    }
+  }
+
+  @action onSubmit(payload, evt) {
+    evt.preventDefault();
+    this.error = '';
+    this.enableReplication.perform(this.args.replicationMode, this.data.mode, payload);
+  }
+}
+
+class EnablePayload {
```

Comment:
Did you make this a class so it could have tracked properties? Since an object key's values are not reliably track-able?

Reply:
Yup! At first I had the tracked properties directly on the component class, but then reading through [this guide](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/#toc_tracked-properties-in-custom-classes) I realized it could be its own separate data class

Reply:
Neat!